projects/OG-Analytics/src/main/java/com/opengamma/analytics/math/surface/FastGridInterpolatedDoublesSurface.java

```diff
@@ -0,0 +1,192 @@
+import com.opengamma.util.tuple.Pair;
+
+// NOTE kirk 2016-03-01 -- This class is *INTENTIONALLY* not following current
+// best practices for generalization or code structure. It is 100% optimized for
+// performance in the case of running large numbers of historical simulations
+// on option pricing. DO NOT attempt to make this class more stylistically in line
+// with other code without carefully understanding and measuring the performance
+// impact of what you're doing.
+// It is *COMPLETELY INTENTIONAL* that there is not a single collection or
+// primitive wrapper in this class. Do not add either.
+
+/**
+ * A fast implementation of {@code Surface} designed specifically for use in grid-based
+ * volatilty surfaces that commonly come up in exchange traded derivatives contexts.
+ * It is optimized for performance over generality, and avoids unnecessary data structure
+ * copies or mutations.
+ * It has a relatively expensive construction cost to allow fast interpolation
+ * and transformation.
+ */
+public final class FastGridInterpolatedDoublesSurface extends Surface<Double, Double, Double> {
+  private static final long serialVersionUID = 1L;
+  private final int _size;
+  private final double[] _xvalues;
+  private final Interpolator1DDataBundle[] _yzBundles;
+  private final Interpolator1D _xInterpolator;
+  private final Interpolator1D _yInterpolator;
+  
+  private FastGridInterpolatedDoublesSurface(
+      int size,
+      double[] xvalues, Interpolator1DDataBundle[] yzBundles,
+      Interpolator1D xInterpolator, Interpolator1D yInterpolator,
+      String name) {
+    super(name);
+    // Intentionally we don't do ArgumentChecker here. It's done in the factory methods.
+    _size = size;
+    _xvalues = xvalues;
+    _yzBundles = yzBundles;
+    _xInterpolator = xInterpolator;
+    _yInterpolator = yInterpolator;
+  }      
+  
+  public static FastGridInterpolatedDoublesSurface fromUnsortedNoCopy(
+      double[] xdata, double[] ydata, double[] zdata,
+      Interpolator1D xInterpolator, Interpolator1D yInterpolator,
+      String name) {
+    ArgumentChecker.notNull(xdata, "xdata");
+    ArgumentChecker.notNull(ydata, "ydata");
+    ArgumentChecker.notNull(zdata, "zdata");
+    ArgumentChecker.notNull(xInterpolator, "xInterpolator");
+    ArgumentChecker.notNull(yInterpolator, "yInterpolator");
+    ArgumentChecker.isTrue(xdata.length > 0, "Must have non-zero size");
+    ArgumentChecker.isTrue(xdata.length == ydata.length, "xdata and ydata must be same size");
+    ArgumentChecker.isTrue(xdata.length == zdata.length, "xdata and zdata must be same size");
+    
+    // NOTE -- This method would be so much easier with built-in tuples.....
+    
+    // First, order xyz triplet by x
+    ParallelArrayBinarySort.parallelBinarySort(xdata, ydata, zdata);
+    
+    // First pass through xdata: just check the cardinality of x and yz data sets
+    double currentX = xdata[0];
+    int numXValues = 0;
```

Comment:
Might be clearer to start this at 1 (guaranteed by the arg checker), avoid the increment at the end, and also initialise `i` to 1 to avoid an unnecessary iteration.